internal/server/iito.go

```diff
@@ -15,3 +23,84 @@ func (s *SOMASServer) runIITOEndOfTurn() error {
 	// TODO:- IITO team
 	return nil
 }
+
+func (s *SOMASServer) runGiftSession() error {
+	s.logf("start runGiftSession")
+	defer s.logf("finish runGiftSession")
+
+	giftRequestDict := s.getGiftRequests()
+	giftOffersDict, err := s.getGiftOffers(giftRequestDict)
+	if err != nil {
+		return err
+	}
+	giftHistoryDict, err := s.getGiftAcceptance(giftOffersDict)
+	if err != nil {
+		return err
+	}
+	err = s.distributeGiftHistory(giftHistoryDict)
+	if err != nil {
+		return err
+	}
+	// Process actions
+	for key, value := range giftHistoryDict {
+		s.logf("Gifts from %s: %v\n", key, value)
+	}
+	return nil
+}
+
+func (s *SOMASServer) getGiftRequests() shared.GiftDict {
+	giftRequestDict := shared.GiftDict{}
+	for id, client := range s.clientMap {
+		giftRequestDict[id] = client.RequestGift()
+	}
+	return giftRequestDict
+}
+func (s *SOMASServer) getGiftOffers(giftRequestDict shared.GiftDict) (map[shared.ClientID]shared.GiftDict, error) {
+	giftOfferDict := map[shared.ClientID]shared.GiftDict{}
+	var err error
+	for id, client := range s.clientMap {
+		giftOfferDict[id], err = client.OfferGifts(giftRequestDict)
+		if err != nil {
+			return giftOfferDict, err
+		}
+	}
+	return giftOfferDict, nil
+}
+func (s *SOMASServer) getGiftAcceptance(giftOffersDict map[shared.ClientID]shared.GiftDict) (map[shared.ClientID]shared.GiftInfoDict, error) {
+	acceptedGifts := map[shared.ClientID]shared.GiftInfoDict{}
+	var err error
+
+	receivedByClientDict := make(map[shared.ClientID]shared.GiftDict)
+
+	//puts all the gifts received by a certain client accesible by the id of that client
+	for idsend := range giftOffersDict {
+		for idto := range giftOffersDict {
+			if idsend == idto {
+				continue
+			}
+			if receivedByClientDict[idsend] == nil {
+				receivedByClientDict[idsend] = make(shared.GiftDict)
+			}
+			receivedByClientDict[idsend][idto] = giftOffersDict[idto][idsend]
+		}
+	}
+
+	for id, client := range s.clientMap {
+		acceptedGifts[id], err = client.AcceptGifts(receivedByClientDict[id])
+		if err != nil {
+			return acceptedGifts, err
+		}
+	}
+	return acceptedGifts, nil
+}
+
+func (s *SOMASServer) distributeGiftHistory(acceptedGifts map[shared.ClientID]shared.GiftInfoDict) error {
+	//Process acceptedGifts
+	for _, client := range s.clientMap {
+		err := client.UpdateGiftInfo(acceptedGifts)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
```

Comment:
Seems good, but the logic gets a bit messy here with nested maps. Still a great implementation, but I'd feel safer if we had some tests?

Reply:
![image](https://user-images.githubusercontent.com/33488131/102904303-76c7ea00-4469-11eb-92fe-3471e5968113.png)

Reply:
Yeah testing is incoming I promise hahaha